### PR TITLE
Add rule S6265: Granting access to S3 buckets to all or authenticated users is security-sensitive

### LINKED ISSUE #3198: Rule S6265: Granting access to S3 buckets to all or authenticated users is security-sensitive
## WHAT
The rule detects when an S3 bucket is configured to be accessible publicly using the JavaScript AWS CDK  library. It does not cover custom policy statement definitions. This has been previously implemented for Python, see [SONARPY-1011](https://github.com/SonarSource/sonar-python/pull/1129).

Specification ticket: [APPSEC-39](https://jira.sonarsource.com/browse/APPSEC-39)
[MMF-2812](https://jira.sonarsource.com/browse/MMF-2812)

## HOW
Detection logic

The rule should detect the following patterns:

1. When `Bucket` constructor is called with an improper configuration
2. When `Bucket.grantPublicAccess` is called
3. When `BucketDeployment` constructor is called with an improper configuration

### `Bucket` constructor

- Method: `aws-cdk-lib.aws_s3.Bucket` constructor
- Argument: `props` 3rd positional argument
- Sensitive properties: 
   - `accessControl`
      - Sensitive values: `PUBLIC_READ`, `PUBLIC_READ_WRITE`, `AUTHENTICATED_READ`
   - `publicReadAccess`
      - Sensitive value: `true` 

The rule raises an issue when a`Bucket` is constructed with one of the above properties set to the incorrect value.
```javascript
new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant',
    accessControl: s3.BucketAccessControl.PUBLIC_READ_WRITE // Sensitive
);

const publicReadAccess = true; // Secondary Location
new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant',
    publicReadAccess: publicReadAccess // Sensitive
);
```

### `Bucket.grantPublicAccess` method

Method: `aws-cdk-lib.aws_s3.Bucket.grantPublicAccess`

Rule raises on the Bucket construct class when _grantPublicAccess_ method is called.
```javascript
const bucketNoncompliant = new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant'
);

bucketNoncompliant.grantPublicAccess(); // Sensitive
```

### `BucketDeployment` constructor

- Method: `aws-cdk-lib.aws_s3.BucketDeployment` constructor
- Argument: `props` 3rd positional argument
- Sensitive propertie:`accessControl`
   - Sensitive values: `PUBLIC_READ`, `PUBLIC_READ_WRITE`, `AUTHENTICATED_READ`

The rule raises an issue when a `BucketDeployment` is contructed with the above propertie set to the incorrect value.
```javascript
const bucketNoncompliant = new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant'
);

new s3deploy.BucketDeployment(this, 'upload', {
    accessControl: s3.BucketAccessControl.PUBLIC_READ_WRITE // Sensitive
});
```

## Code examples

https://github.com/SonarSource/security-expected-issues/pull/721

## Message and highlighting

See https://github.com/SonarSource/rspec/pull/1084

### PR DESCRIPTION
**Pull request by ilia-kebets-sonarsource**
Fixes #3198

### CODE REVIEW

File: eslint-bridge/src/rules/aws-s3-bucket-access.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+const ACCESS_CONTROL_KEY = 'accessControl';
+const INVALID_ACCESS_CONTROL_VALUES = ['PUBLIC_READ', 'PUBLIC_READ_WRITE', 'AUTHENTICATED_READ'];
+
+const PUBLIC_READ_ACCESS_KEY = 'publicReadAccess';
+const INVALID_PUBLIC_READ_ACCESS_VALUE = true;
+
+export const rule: Rule.RuleModule = {
+  create(context: Rule.RuleContext) {
+    return mergeRules(
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/rules/aws-s3-bucket-access.ts:53]: I wonder whether we could generalize this function to merge any number of rules. What do you think?

File: eslint-bridge/src/rules/aws-s3-bucket-access.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  bucketConstructor: estree.NewExpression,
+  propName: string,
+  propValue: boolean,
+) {
+  const property = getProperty(context, bucketConstructor, propName);
+  if (property == null) {
+    return;
+  }
+  const propertyLiteralValue = getValueOfExpression(context, property.value, 'Literal');
+  if (propertyLiteralValue !== undefined && propertyLiteralValue.value === propValue) {
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/rules/aws-s3-bucket-access.ts:111]: What about optional chaining ?

File: eslint-bridge/src/utils/s3-rule-template.ts [hunk 3]
```diff
@@ -49,24 +49,38 @@ export function S3BucketTemplate(
     create(context: Rule.RuleContext) {
       return {
         NewExpression: (node: estree.NewExpression) => {
-          if (isS3BucketConstructor(node, context)) {
+          if (isS3BucketConstructor(context, node)) {
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/utils/s3-rule-template.ts:52]: Out of curiosity, why inverting them?

File: its/sources/file-for-rules/S6265.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+const s3 = require('aws-cdk-lib/aws-s3');
+
```
Comment (yassin-kammoun-sonarsource) [inline its/sources/file-for-rules/S6265.js:2]: Let's add a compliant example as well.

**Review by yassin-kammoun-sonarsource — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/utils/s3-rule-template.ts (hunk 3 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/src/utils/s3-rule-template.ts:52]: I found that it's a general pattern of our methods to have context first.

File: eslint-bridge/src/rules/aws-s3-bucket-access.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
+/*
```
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/src/rules/aws-s3-bucket-access.ts]: done

File: eslint-bridge/src/rules/aws-s3-bucket-access.ts (hunk 1 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/src/rules/aws-s3-bucket-access.ts]: Done thanks to @victor-diez-sonarsource !

File: its/sources/file-for-rules/S6265.js (hunk 4 above)
Comment (ilia-kebets-sonarsource) [reply its/sources/file-for-rules/S6265.js:2]: done

File: eslint-bridge/src/rules/aws-s3-bucket-access.ts (hunk 2 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/src/rules/aws-s3-bucket-access.ts]: somehow, this has decreased the coverage..

File: eslint-bridge/src/utils/s3-rule-template.ts (hunk 3 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/utils/s3-rule-template.ts:52]: 👍

**Review by yassin-kammoun-sonarsource — APPROVED** [review]

**Pull request merged**